lib/MetadataStore.js

```diff
@@ -0,0 +1,385 @@
+        } catch (e) {
+          Cu.reportError("MetadataStore failed to insert metadata: " + e.message);
+          throw e;
+        }
+
+        /* 2. inserts the favicon into page_images, handles the special case, in which
+         * the favicon already exsits */
+        let faviconBindings = this._getFaviconParameters(row);
+        try {
+          yield this._conn.executeCached(SQL_INSERT_IMAGES, faviconBindings);
+          image_ids.push(yield this._getLastInsertRowID());
+        } catch (e) {
+          try {
+            image_ids.push(yield this._getImageIDByURL(faviconBindings.url));
+          } catch (e) {
+            Cu.reportError("MetadataStore failed to insert favicon: " + e.message);
+            throw e; /* force this transaction to rollback */
+          }
+        }
+
+        /* 3. inserts all the preview images */
+        for (let image of row.images) {
+          let imageBindings = this._getImageParameters(image);
+          try {
+            yield this._conn.executeCached(SQL_INSERT_IMAGES, imageBindings);
+            image_ids.push(yield this._getLastInsertRowID());
+          } catch (e) {
+            try {
+              image_ids.push(yield this._getImageIDByURL(imageBindings.url));
+            } catch (e) {
+              Cu.reportError("MetadataStore failed to fetch the id of favicon: " + e.message);
+              throw e; /* force this transaction to rollback */
+            }
+          }
+        }
+
+        /* 4. inserts relations into the page_metadata_images */
+        for (let image_id of image_ids) {
+          yield this._conn.executeCached(SQL_INSERT_METADATA_IMAGES, [metadata_id, image_id]);
+        }
+      }.bind(this));
+    }
+  }),
+
+  /**
+   * Drops all the tables and the corresponding indices, the table schema remains
+   *
+   * Returns a promise that is resolved upon success, or rejected if an exception occurs
+   */
+  asyncDrop: Task.async(function*() {
+    try {
+      yield this._conn.executeTransaction(function*() {
+        for (let drop of SQL_DROPS) {
+          yield this._conn.execute(drop);
+        }
+      }.bind(this));
+      yield this._createTableSchema();
+    } catch (e) {
+      Cu.reportError("MetadataStore failed to drop: " + e.message);
+      throw e;
+    }
+  }),
+
+  /**
+   * Executes arbitrary query against metadata database. For bulk insert, use
+   * asyncInsert function instead
+   *
+   * @param {String} aSql
+   *        SQL query to execute
+   * @param {Object} [optional] aOptions
+   *        aOptions.columns - an array of column names. if supplied the return
+   *        items will consists of objects keyed on column names. Otherwise
+   *        array of raw values is returned in the select order
+   *        aOptions.param - an object of SQL binding parameters
+   *        aOptions.callback - a callback to handle query raws
+   *
+   * Returns a promise with the array of retrieved items
+   */
+  executeQuery(aSql, aOptions = {}) {
```

Comment:
nit: use `Task.async`. The function with a similar name in `PlacesProvider` could use some refactoring too